prov/shm/src/smr_progress.c

```diff
@@ -241,16 +249,29 @@ static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
 		FI_WARN(&smr_prov, FI_LOG_AV, "mmap error %s\n", strerror(errno));
 		return errno;
 	}
-	ret = ofi_copy_to_iov(iov, iov_count, 0, mapped_ptr, cmd->msg.hdr.size);
-	if (ret != cmd->msg.hdr.size) {
-		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-			"mmap iov copy out error\n");
-		return FI_EIO;
+	if (cmd->msg.hdr.op == ofi_op_read_req) {
+		*total_len = ofi_total_iov_len(iov, iov_count);
+		ret = ofi_copy_from_iov(mapped_ptr, *total_len, iov, iov_count, 0);
+		if (ret != *total_len) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy in error\n");
+			return FI_EIO;
+		}
+	} else {
+		ret = ofi_copy_to_iov(iov, iov_count, 0, mapped_ptr,
+				      cmd->msg.hdr.size);
+		if (ret != cmd->msg.hdr.size) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy out error\n");
+			return FI_EIO;
+		}
+		*total_len = ret;
```

Comment:
This isn't needed, as total_len is set 2 lines down.

Reply:
I removed the setting 2 lines down.